Fix errors with libraries when old mongo is default (first) modulestore (SOL-234)

bradenmacdonald opened the pull request:
**Background**: This commit fixes a major error that went unnoticed during the content libraries MVP: If an edX instance is configured such that old mongo is the default modulestore in MixedModulestore's options (regardless of the value of `DEFAULT_STORE_FOR_NEW_COURSE`), then content libraries cannot be edited in Studio.

**Explanation**: Libraries are always created in split, but once a library has been created, the system would only ever look for libraries in the default modulestore. Thus it wouldn't be able to actually read the contents of the newly created library. This fixes the problem by searching all modulestores for the requested library in the same manner we do for courses.

**JIRA Ticket**: [SOL-234](https://openedx.atlassian.net/browse/SOL-234)

**Discussions**: Error was pointed out by @andy-armstrong and @antoviaque. Fix has not yet been discussed.

**Dependencies**: None

**Sandbox URL**: TBD

**Partner information**: 3rd party-hosted open edX instance, for an edX solutions client.

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=6657&repo=edx%2Fedx-platform

bradenmacdonald commented:
Test failure is unrelated.

bradenmacdonald commented:
@andy-armstrong this should fix the problem you told me about by email. Would you be able to review it?

@dmitchell if you happen to have time before you go, can you please take a quick look at this or let me know who else can review this?

andy-armstrong commented on common/lib/xmodule/xmodule/modulestore/mixed.py:
```diff
@@ -172,8 +172,12 @@ def _get_modulestore_for_courseid(self, course_id=None):
             if mapping is not None:
                 return mapping
             else:
+                if isinstance(course_id, LibraryLocator):
```
It seems a shame to have this logic hidden inside this method. I would move this logic into ``has_course`` and possibly rename it since it isn't necessarily only a course that the store has. Otherwise this sort of tricky logic will be required in other places too, I would imagine.

If it is the wrong thing for ``has_course`` to return true in this case, then I'd argue that it should throw an exception if the locator it receives isn't for a course (if it doesn't already).

bradenmacdonald replied:
This method, `_get_modulestore_for_courseid` is key to how MixedModulestore works and is used by almost every other method in this file, so it would create a ton of code redundancy to move this branching logic anywhere else I think. It would be more clear if it were called `_get_modulestore_for_course_or_library` but that's too long and this is one of just many cases where the code that uses it (methods like `get_item`, `update_item` etc.) doesn't care if they're dealing with courses or libraries and shouldn't have to know about the difference.

andy-armstrong replied:
Sorry, that wasn't clear. My point is that instead of adding this logic into this method, I would move it into ``has_course`` so that the call below doesn't have to be changed. It doesn't make sense to me that only this method knows how to determine if a library is managed by a modulestore.

Part of the problem is that the parameter passed in is called ``course_id``, but it is no longer a course_id in the true meaning of the term. Is a library a course or not? If not, we need a term for a 'thing that is either a library or a course'. If a library is a course, then this logic unambiguously should move into store.has_course.

bradenmacdonald replied:
I see what you're saying now - sorry I misunderstood. There are many different `has_course` methods...

A library is not a course, though they have similar properties. A library ID (key) *is* a subclass of *CourseKey*, the most generic key for courses, but `LibraryLocator` is not a subclass of `CourseLocator` and the concept of a library is different from that of a course.

The reason I didn't put this code into `store.has_course` is that `store.get_course()` and `store.get_library()` are already separated, and `get_course()` will never return a library. So it doesn't make sense to me if I could call `store.has_course(library_id)` and get `True` but then call `store.get_course(library_id)` and get an exception.

doctoryes replied:
Now that a `course_id` can refer to `LibraryLocator` or `CourseLocator` since we're storing both courses and problem libraries in the modulestores as separate entities, it's unfortunate that we still use `course_id` to refer to both types of locators - it's confusing. Though - I can't think of a better term at the moment.

mattdrayer replied:
How about 'locator_id'?

doctoryes replied:
`locator_id` is a bit redundant - maybe `locator`?

bradenmacdonald replied:
LibraryLocator *is* a type of `CourseKey` at the moment (though I'd like that to change in the long term), so technically a course_key does not imply not a library key, here or elsewhere.

(If this method was called "_get_modulestore_for_course" it would be wrong because a course is not a library - but a course ID can be a library ID)

However, I don't like relying on technicalities and would prefer to rename this method and its argument for clarity. In split I used "courselike" to refer to "course or library". But see my other comment below just now - I think we need to define a standard term and use it consistently.

Please let me know your thoughts and what specifically you'd like changed before we can merge this PR.

doctoryes replied:
Correct - all locators are currently `CourseKey`s. The problem is that `CourseKey

bradenmacdonald replied:
Yes, agreed about the need for a new base key type - that [was always my plan](https://openedx.atlassian.net/wiki/display/SOL/Content+Libraries#ContentLibraries-NewLocators).

I like `_get_modulestore_for_courselike(locator)` for now - I can make that change.

bradenmacdonald replied:
Renamed in 044dc7fb6b15608d72125c8335ee2551e4f5dfae

andy-armstrong commented:
@bradenmacdonald I'm not really a modulestore guy, so I recommend that you ask for a review from the platform team (@cpennington or @doctoryes).

andy-armstrong commented:
@bradenmacdonald Thanks for tracking this down, by the way.  This does indeed look like it is the cause of the bug I ran into.

bradenmacdonald commented:
@cpennington Could you please give this a quick review?

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=6657&repo=edx%2Fedx-platform

cpennington commented:
Between the library/course and xblock/aside divisions, we have a proliferation of methods with similar names in modulestore and xblock (`get_course/get_library`, for instance). I don't really have a good answer on how to fix it, but it would be nice to figure out a way.

doctoryes commented:
How about a pattern that defines a `get_item` method (or `get_store_item` or `get_entity`) and then introspects the passed-in locator or id to infer for what it should be looking (course/library)? This PR essentially implements that logic but without changing the top-level API.

bradenmacdonald commented:
To  be clear: In MixedModuleStore and its friends, there are three families of methods that distinguish between courses and libraries: `get_course`/`get_library`, `has_course`/`has_library`, and `create_course`/`create_library`. All of the other (block-level) methods don't care about libraries vs. courses. I think the separation we have now at this level is important, despite the obvious code similarities, because libraries and courses do have different properties and can't be treated the same way in general.

What I think would be good to see long-term is the creation of a new concept that encapsulates courses, libraries, and future smaller re-usable "units" of related XBlocks. This could be called ContentPackage, Capsule, StructuredContent, BlockContainer, ContentRoot, Courselike, or ???. Then we could guarantee a few common properties of these content tree objects - they have a root node, they have children, they have a name, etc. Then code that relies on just that subset of properties can use new methods safely without worrying about libraries vs. courses.

doctoryes commented:
:+1:  after tests pass.

bradenmacdonald commented:
Thanks @doctoryes

bradenmacdonald commented:
Tests complete and the only failure is a flaky bok-choy video test. @cdodge @mattdrayer could one of you please merge if this good to go now?

antoviaque commented:
@bradenmacdonald I think it's better if you merge it yourself - this way you will be the one listed for verification of the change during the weekly release process.

However it would need a second thumb up to be good to go, yes, which @cdodge or @mattdrayer could provide - not sure how comfortable you are with modulestore code though? @e-kolpakov it might make sense for you to do that second review, if you feel comfortable doing it.

e-kolpakov commented on common/lib/xmodule/xmodule/modulestore/mixed.py:
```diff
@@ -157,35 +157,39 @@ def __init__(
                 for store in self.modulestores:
-                    if store.has_course(course_id):
-                        self.mappings[course_id] = store
+                    if has_course(store):
```
@bradenmacdonald nit: something like `has_courselike(store, locator)` would be more self-explaining here, so you might want changing the name and signature of this lambda.

e-kolpakov commented:
@bradenmacdonald other than a minor naming ntpick, loogs good :+1:

bradenmacdonald commented:
Renamed variable mentioned by Eugeny and squashed to one commit. If there are no other comments and the tests pass, I'll merge this later today. Thanks everyone.
